Disallow annotations on instance attributes outside __init__

andrewor14 opened the pull request:
**Summary**: This commit solves the first part of https://github.com/pytorch/pytorch/issues/52306, which disallows type annotations on instance attributes inside any method other than the constructor.

**Test Plan**: Added test to test_types.py.

**Reviewers**: Zhengxu Chen

**Subscribers**: Zhengxu Chen, Yanan Cao, Peng Wu, Yining Lu

**Tasks**: T103941984

**Tags**: pytorch

**Fixes** #52306

zhxchen17 reviewed (commented):
Overall looks great! @andrewor14 thanks for digging in and contributing to the frontend :) 
Only one concern about code placement.

Commit:
Disallow annotations on instance attributes outside __init__ (#67051)

Summary:
**Summary**: This commit solves the first part of https://github.com/pytorch/pytorch/issues/52306, which disallows type annotations on instance attributes inside any method other than the constructor.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/67051

Test Plan:
Added test to test_types.py.

**Reviewers**: Zhengxu Chen

**Subscribers**: Zhengxu Chen, Yanan Cao, Peng Wu, Yining Lu

**Tasks**: T103941984

**Tags**: pytorch

**Fixes** https://github.com/pytorch/pytorch/issues/52306

Differential Revision: D31843527

Pulled By: andrewor14

fbshipit-source-id: 27a9b854ab7d082ade5bbf91a1816ee9584b0b06
